File: NeetCode/Heap/K_Closest_Points_To_Origin/k__closest__points__to__origin.py
from typing import Optional, List
import heapq, math


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        def dist(x, y):
            return math.sqrt(x**2 + y**2)

        distList = []
        for point in points:
            d = dist(point[0], point[1])
            heapq.heappush(distList, (-d, point))
            if len(distList) > k:
                heapq.heappop(distList)

        return [i[1] for i in distList]


if __name__ == "__main__":
    sol = Solution()
    points = [[0, 2], [2, 0], [2, 2]]
    k = 2

    print(sol.kClosest(points, k))
    print("Running Solution...")

File: NeetCode/Heap/K_Closest_Points_To_Origin/K_Closest_Points_To_Origin.cpp
// K_Closest_Points_To_Origin.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> kClosest(vector<vector<int>> &points, int k)
    {
        priority_queue<pair<int, vector<int>>> distList;
        for (const vector<int> &point : points)
        {
            int dist = point[0] * point[0] + point[1] * point[1];
            pair<int, vector<int>> tmp = {dist, point};
            distList.push(tmp);
            if (distList.size() > k)
            {
                distList.pop();
            }
        }

        vector<vector<int>> results;
        while (!distList.empty())
        {
            results.push_back(distList.top().second);
            distList.pop();
        }
        return results;
    }
};
